src/btree/bt_read.c

```diff
@@ -407,11 +407,13 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
                 ++force_attempts;
                 ret = __wt_page_release_evict(session, ref, 0);
                 /*
-                 * If forced eviction succeeded, don't retry. If it failed, stall.
+                 * If forced eviction succeeded or if it failed because other session doing
+                 * checkpoint, don't retry. Else if it failed, stall.
                  */
-                if (ret == 0)
+                if (ret == 0 || __wt_btree_syncing_by_other_session(session)) {
```

Comment:
This change will indeed allow for faster throughput while a checkpoint is happening. The downside is that it will allow for pages to grow much bigger, which will generally result in a bigger disruption after the checkpoint completes - when exclusive access is acquired and held for longer to reconcile the now very large page.

The current code doesn't explain that trade off very well, but it is done deliberately. The throttling in place here avoids a bigger disruption after the checkpoint completes.

If you think this change is better, please include a test workload along with the PR that demonstrates the benefit, and also look for the post-checkpoint reconciliation consequence.

Reply:
hi, @agorrod 
We have this type of user, and the business scenarios are as follows:
Due to Redis not supporting some complex queries, they use MongoDB to replace MySQL+Redis. Their total data volume is not large, but their find requests are high, update and insert write requests are not high, and they are sensitive to latency.

But I found a problem that when a node performs a checkpoint operation, the delay will increase significantly. By analyzing the diagnostic data of diagnose.data, it is confirmed that the main reason is force retry. When used as a checkpoint, the diagnostic value of page_forcible_evictblocked is very high.

For this type of business, users are sensitive to read and write latency, so the requested mongodb cluster size will be relatively large, such as 32CPU/128G MEM/200G disk, where almost all data can be cached in memory.

Due to the low write request, the growth of the page is a slow process, and even if evict is not done during the checkpoint, the page will not change quickly.

Due to the large size of the memory specifications, there is basically no triggering of evication_dirty_target=5 and evication_target=80. Therefore, all evict+reconciles are completed by the user thread through the __wt_page_release_evict().

Because the update operation is not very high, the page will not undergo significant changes in the short term. Therefore, I have some optimization ideas, and I am not sure if there will be significant returns:
1. Add a configuration. If this configuration is enabled, do not retry if the return value is busy, otherwise retry is required.

I conducted a simple test and found that stacked's sleep time is more time-consuming than big page's normal evict.

This way, we can avoid a bigger disruption by the user thread's __wt_page_release_evict When the checkpoint is completed, the next time data is written to this page, normal evict operations can be performed.

thanks.

Reply:
Thanks for the context - generally we prefer to keep this type of information in our Jira tickets. It is more accessible and remains searchable in the future when we come to revisit why we made decisions.

As such I've moved the content into the Jira ticket and we will follow up there.